qutebrowser/browser/greasemonkey.py

```diff
@@ -0,0 +1,289 @@
+            scriptName=self._name,
+            scriptInfo=self.meta_json() or 'null',
+            scriptMeta=self.meta_raw() or 'null')
+        return '\n'.join(
+            ["(function(){", gm_bootstrap, self._code, "}).call();"])
+
+    def meta_json(self):
+        return json.dumps({
+            'name': self._name,
+            'description': self._description,
+            'matches': self._includes,
+            'includes': self._includes,
+            'excludes': self._excludes,
+            'run-at': self._run_at,
+        })
+
+    def meta_raw(self):
+        pass
+
+
+class GreasemonkeyManager(QObject):
+
+    """Manager of userscripts and a greasemonkey compatible environment.
+
+    Signals:
+        scripts_reloaded: Emitted when scripts are reloaded from disk.
+            Any any cached or already-injected scripts should be
+            considered obselete.
+    """
+
+    scripts_reloaded = pyqtSignal()
+
+    def __init__(self):
+        super().__init__()
+        self.load_scripts()
+
+    @cmdutils.register(name='greasemonkey-reload',
+                       instance='greasemonkey')
+    def load_scripts(self):
+        """Re-Read greasemonkey scripts from disk."""
+        self._run_start = []
+        self._run_end = []
+        self._run_idle = []
+
+        scripts_dir = QDir(_scripts_dir())
+        log.greasemonkey.debug("Reading scripts from: %s",
+                               scripts_dir.absolutePath())
+        for script_filename in scripts_dir.entryList(['*.js'], QDir.Files):
+            script_path = scripts_dir.absoluteFilePath(script_filename)
+            log.greasemonkey.debug("Trying to load script: %s", script_path)
+            with open(script_path, encoding='utf-8') as script_file:
+                try:
+                    script = GreasemonkeyScript.parse(script_file.read())
+                except Exception as err:
+                    log.greasemonkey.warning("Unable to load: %s %s",
+                                             script_path, err)
+                    # Catch-all :/
+                    continue
+                if script.run_at() == 'document-start':
+                    self._run_start.append(script)
+                elif script.run_at() == 'document-end':
+                    self._run_end.append(script)
+                elif script.run_at() == 'document-idle':
+                    self._run_idle.append(script)
+                else:
+                    log.greasemonkey.warning(("Script %s has invalid run-at "
+                                              "defined, defaulting to "
+                                              "document-end"), script_path)
+                    # Default as per
+                    # https://wiki.greasespot.net/Metadata_Block#.40run-at
+                    self._run_end.append(script)
+                log.greasemonkey.debug("Loaded script: %s", script.name())
+        self.scripts_reloaded.emit()
+
+    def scripts_for(self, url):
```

Comment:
Looking at https://wiki.greasespot.net/Include_and_exclude_rules this should probably also limit the list of schemes those scripts apply to?

Reply:
Phew, you sure do do your research. I kind of want to include file:// too since apparently it is included to stop malicious scripts reading your local files. Which is silly because greasemonkey scripts are privileged anyway. It's not like random sites are going to be installing them without you knowing. Sure people can be social engineered into installing a script which does bad stuff but getting them to run an exe as administrator is probably less complicated.

Reply:
You're right - I didn't really look into what the issue was when writing the comment, but now that I did, I'll have to agree.